### PR TITLE
Clarify what to do when ALPN negotiation fails

### PR DESCRIPTION
**Pull request by ghedo**
Just so you don't have to read the whole RFC7301.

### CODE REVIEW

**Comment by ghedo** [comment]
Yeah I was thinking about that. I wonder if it would make sense for the client to simply send PROTOCOL_VIOLATION instead of a TLS alert.

**Comment by ghedo** [comment]
@martinthomson I added a clarification to allow clients to send no_application_negotiated as well. Should I send an email to the mailing list for discussion now that the new process is in effect?

**Comment by martinthomson** [comment]
Unfortunately yes.  This is a design change so it will need an issue as well.

I like it and we should do it, but can you add the error code we would use in CONNECTION_CLOSE to the text?

**Comment by ghedo** [comment]
@martinthomson ok, added error code, created issue (#2495) and sent email to the list.

File: draft-ietf-quic-tls.md [hunk 1]
```diff
@@ -1216,7 +1216,9 @@ negotiation.  TLS uses Application Layer Protocol Negotiation (ALPN)
 {{!RFC7301}} to select an application protocol.  Unless another mechanism is
 used for agreeing on an application protocol, endpoints MUST use ALPN for this
 purpose.  When using ALPN, endpoints MUST abort a connection if an application
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md]: This isn't within your change, but what  does "abort" mean. Should this be "drop QUIC connection state"?

Comment (martinthomson) [reply draft-ietf-quic-tls.md]: I think that an immediate close is what was intended.

File: draft-ietf-quic-tls.md [hunk 2]
```diff
@@ -1216,7 +1216,9 @@ negotiation.  TLS uses Application Layer Protocol Negotiation (ALPN)
...
-protocol is not negotiated.
+protocol is not negotiated with a no_application_protocol alert (error code
+0x178). While {{!RFC7301}} only specifies that servers use this alert, QUIC
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md]:
```suggestion
0x178, see {{tls-errors}}).  While {{!RFC7301}} only specifies that servers use this alert, QUIC
```

File: draft-ietf-quic-tls.md [hunk 3]
```diff
@@ -1216,7 +1216,9 @@ negotiation.  TLS uses Application Layer Protocol Negotiation (ALPN)
 {{!RFC7301}} to select an application protocol.  Unless another mechanism is
 used for agreeing on an application protocol, endpoints MUST use ALPN for this
 purpose.  When using ALPN, endpoints MUST abort a connection if an application
-protocol is not negotiated.
+protocol is not negotiated with a no_application_protocol alert (error code
```
Comment (janaiyengar) [inline draft-ietf-quic-tls.md]:
```suggestion
protocol is not negotiated with a no_application_protocol TLS alert (QUIC error code
```

File: draft-ietf-quic-tls.md (hunk 1 above)
Comment (janaiyengar) [reply draft-ietf-quic-tls.md]:
```suggestion
purpose.  When using ALPN, endpoints MUST immediately close a connection (see Section 10.3 in {{QUIC-TRANSPORT}}) if an application
```

**Comment by ghedo** [comment]
@janaiyengar thanks, suggestions accepted.

**Pull request merged**